### PR TITLE
Added our own code to parse a large report from Google

### LINKED ISSUE #929: User stats is failing each morning
rake scholarsphere:stats:user_stats exited with a non-zero status. WARNING: Use strings for Figaro configuration. false was converted to "false". DEPRECATION WARNING: The configuration option `config.serve_static_assets` has been renamed to `config.serve_static_files` to clarify its role (it merely enables serving everything in the `public` folder and is unrelated to the asset pipeline). The `serve_static_assets` alias will be removed in Rails 5.0. Please migrate your configuration files accordingly. (called from block in <top (required)> at /opt/heracles/deploy/scholarsphere/releases/20170605152656/config/environments/production.rb:50) rake aborted! NoMethodError: undefined method `each' for true:TrueClass /opt/heracles/deploy/scholarsphere/shared/bundle/ruby/2.3.0/bundler/gems/sufia-603e8fc8bd4f/app/services/sufia/user_stat_importer.rb:120:in `tally_results' /opt/heracles/deploy/scholarsphere/shared/bundle/ruby/2.3.0/bundler/gems/sufia-603e8fc8bd4f/app/services/sufia/user_stat_importer.rb:53:in `block in process_files' /opt/heracles/deploy/scholarsphere/shared/bundle/ruby/2.3.0/bundler/gems/sufia-603e8fc8bd4f/app/services/sufia/user_stat_importer.rb:50:in `each' /opt/heracles/deploy/scholarsphere/shared/bundle/ruby/2.3.0/bundler/gems/sufia-603e8fc8bd4f/app/services/sufia/user_stat_importer.rb:50:in `process_files' /opt/heracles/deploy/scholarsphere/shared/bundle/ruby/2.3.0/bundler/gems/sufia-603e8fc8bd4f/app/services/sufia/user_stat_importer.rb:30:in `block in import' /opt/heracles/deploy/scholarsphere/shared/bundle/ruby/2.3.0/bundler/gems/sufia-603e8fc8bd4f/app/services/sufia/user_stat_importer.rb:23:in `each' /opt/heracles/deploy/scholarsphere/shared/bundle/ruby/2.3.0/bundler/gems/sufia-603e8fc8bd4f/app/services/sufia/user_stat_importer.rb:23:in `import' /opt/heracles/deploy/scholarsphere/shared/bundle/ruby/2.3.0/bundler/gems/sufia-603e8fc8bd4f/tasks/stats_tasks.rake:6:in `block (3 levels) in <top (required)>' /opt/heracles/deploy/scholarsphere/shared/bundle/ruby/2.3.0/gems/rake-12.0.0/exe/rake:27:in `<top (required)>' /opt/heracles/deploy/.rbenv/versions/2.3.3/bin/bundle:22:in `load' /opt/heracles/deploy/.rbenv/versions/2.3.3/bin/bundle:22:in `<main>' Tasks: TOP => sufia:stats:user_stats (See full trace by running task with --trace) 2017-06-09 03:00:22 -0400 (26368) Sufia::UserStatImporter: Begin import of User stats. 2017-06-09 03:00:40 -0400 (26368) ActiveFedora: loading fedora config from /opt/heracles/deploy/scholarsphere/releases/20170605152656/config/fedora.yml 2017-06-09 03:00:40 -0400 (26368) ActiveFedora: loading solr config from /opt/heracles/deploy/scholarsphere/releases/20170605152656/config/solr.yml 2017-06-09 03:00:49 -0400 (26368) Attempted to init base path `prod`, but it already exists 2017-06-09 03:00:49 -0400 (26368) Sufia::UserStatImporter: Retried FileViewStat on #<struct Struct::UserRecord id=1720, user_key="yft5050", last_stats_update=Sun, 24 Mar 2013 00:00:00 +0000> for file 12j62s508f too many times. 2017-06-09 03:00:49 -0400 (26368) Sufia::UserStatImporter: Last exception Private key file for Google analytics was expected at '/scholarsphere/config_production_new/scholarsphere/ga-privatekey.p12', but no file was found.

### PR DESCRIPTION
**Pull request by carolyncole**
No longer individual queries for each Work or FileSet

Fixes #929 
refs #1020 becuase with this change we should be able to get at the FileSet downloads for each page when the page loads.

### CODE REVIEW

File: spec/services/download_report_spec.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      allow(report).to receive(:google_analytics_download_results).and_return(download_results)
+      allow(generic_file_with_fileset).to receive(:file_sets).and_return([FileSet.new(id: '3xs55m950_file', depositor: user2.login)])
+      allow(ActiveFedora::Base).to receive(:find).with('3xs55m950').twice.and_return(generic_file_with_fileset)
+      allow(ActiveFedora::Base).to receive(:find).with('3xs55m9505').once.and_return(GenericWork.new(id: '3xs55m9505', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('1vh53wt10').once.and_return(GenericWork.new(id: '1vh53wt10', depositor: user2.login))
+      allow(ActiveFedora::Base).to receive(:find).with('1vh53wt10z').once.and_return(FileSet.new(id: '1vh53wt10z', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('fj67314220').once.and_return(FileSet.new(id: 'fj67314220', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('nzs25x853x').once.and_return(FileSet.new(id: 'nzs25x853x', depositor: user2.login))
+      allow(ActiveFedora::Base).to receive(:find).with('4f4752g26g').once.and_return(FileSet.new(id: '4f4752g26g', depositor: user1.login))
```
Comment (awead) [inline spec/services/download_report_spec.rb]: Where are these other three ids being used in this test?

File: spec/services/download_report_spec.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+       OpenStruct.new(eventCategory: 'Files', eventAction: 'Downloaded', eventLabel: '(not set)', date: '20170820', pagePath: '/downloads/1vh53wt10z', totalEvents: '1')]
+    }
+
+    let(:generic_file_with_fileset) { GenericWork.new(id: '3xs55m950', depositor: user2.login) }
+
+    before do
+      allow(report).to receive(:google_analytics_download_results).and_return(download_results)
+      allow(generic_file_with_fileset).to receive(:file_sets).and_return([FileSet.new(id: '3xs55m950_file', depositor: user2.login)])
+      allow(ActiveFedora::Base).to receive(:find).with('3xs55m950').twice.and_return(generic_file_with_fileset)
+      allow(ActiveFedora::Base).to receive(:find).with('3xs55m9505').once.and_return(GenericWork.new(id: '3xs55m9505', depositor: user1.login))
```
Comment (awead) [inline spec/services/download_report_spec.rb]: Is this id being used?

File: spec/services/download_report_spec.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      allow(ActiveFedora::Base).to receive(:find).with('1vh53wt10').once.and_return(GenericWork.new(id: '1vh53wt10', depositor: user2.login))
```
Comment (awead) [inline spec/services/download_report_spec.rb]: Is this id being used?

File: spec/services/download_report_spec.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      allow(ActiveFedora::Base).to receive(:find).with('1vh53wt10z').once.and_return(FileSet.new(id: '1vh53wt10z', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('fj67314220').once.and_return(FileSet.new(id: 'fj67314220', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('nzs25x853x').once.and_return(FileSet.new(id: 'nzs25x853x', depositor: user2.login))
+      allow(ActiveFedora::Base).to receive(:find).with('4f4752g26g').once.and_return(FileSet.new(id: '4f4752g26g', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('other_missing').and_raise(ActiveFedora::ObjectNotFoundError)
```
Comment (awead) [inline spec/services/download_report_spec.rb]: we're allowing an exception to be raised, but are we testing that exception anywhere?

File: spec/services/user_stats_importer_spec.rb [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      allow(ActiveFedora::Base).to receive(:find).with('3xs55m950').twice.and_return(generic_file_with_fileset)
+      allow(ActiveFedora::Base).to receive(:find).with('3xs55m9505').once.and_return(GenericWork.new(id: '3xs55m9505', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('1vh53wt10').once.and_return(GenericWork.new(id: '1vh53wt10', depositor: user2.login))
+      allow(ActiveFedora::Base).to receive(:find).with('1vh53wt10z').once.and_return(FileSet.new(id: '1vh53wt10z', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('fj67314220').once.and_return(FileSet.new(id: 'fj67314220', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('nzs25x853x').once.and_return(FileSet.new(id: 'nzs25x853x', depositor: user2.login))
+      allow(ActiveFedora::Base).to receive(:find).with('4f4752g26g').once.and_return(FileSet.new(id: '4f4752g26g', depositor: user1.login))
+      allow(ActiveFedora::Base).to receive(:find).with('other_missing').and_raise(ActiveFedora::ObjectNotFoundError)
```
Comment (awead) [inline spec/services/user_stats_importer_spec.rb]: Don't see this used anywhere.

File: app/services/analytics/view_report.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    end
+
+    def file_set_page_views
+      @file_set_page_views ||= google_analytics_view_results.select { |result| result.pagePath.include?('file_sets') }
+    end
+
+    private
+
+      def google_analytics_view_results
+        return @google_analytics_view_results if @google_analytics_view_results.present?
```
Comment (awead) [inline app/services/analytics/view_report.rb]: I took me a little while to figure out you're memoizing this. Is this better?
``` ruby
def google_analytics_view_results
  @google_analytics_view_results ||= init_view_results
end

def init_view_results
  results = Sufia::Pageview.results(Sufia::Analytics.profile, start_date: start_date, end_date: end_date, sort: :pagePath)
  results.dimensions << :pagePath
  results
end
```

File: app/services/analytics/download_report.rb [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    attr_reader :start_date, :end_date
+
+    def initialize(start_date, end_date)
+      @start_date = start_date
+      @end_date = end_date
+    end
+
+    def work_downloads
+      initialize_page_downloads
+      @work_downloads
```
Comment (awead) [inline app/services/analytics/download_report.rb]: Can this be: `@work_downloads ||= initialize_page_downloads`

File: app/services/analytics/download_report.rb [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    end
+
+    def file_set_downloads
+      initialize_page_downloads
+      @file_set_downloads
```
Comment (awead) [inline app/services/analytics/download_report.rb]: same here: `@file_set_downloads ||= initialize_page_downloads`

File: app/services/analytics/download_report.rb [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+        profile = Sufia::Analytics.profile
+        @google_analytics_download_results = Sufia::Download.results(profile, start_date: start_date, end_date: end_date, sort: :pagePath)
+        @google_analytics_download_results.dimensions << :pagePath
+
+        @google_analytics_download_results
+      end
+
+      def initialize_page_downloads
+        return if @work_downloads.present?
```
Comment (awead) [inline app/services/analytics/download_report.rb]: you shouldn't need this is the above change can be made

File: app/services/analytics/download_report.rb [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+        @work_downloads = []
+        @file_set_downloads = []
+        google_analytics_download_results.each do |page_download|
+          @work_downloads, @file_set_downloads = classify_download(page_download, @work_downloads, @file_set_downloads)
```
Comment (awead) [inline app/services/analytics/download_report.rb]: I feel like if you could put `classify_download` within thie each loop, we would need to pass around the `work_downloads` and `file_set_downloads` variables. It's confusing to read.

File: app/services/analytics/download_report.rb [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        puts "Error loading object for download id: #{id}, error: #{e}"
+        return work_downloads, file_set_downloads
+      end
+
+      def path_to_id(page_path)
+        page_path.split('/').last
+      end
+
+      def lookup_object(id)
+        @last_object ||= GenericWork.new
```
Comment (awead) [inline app/services/analytics/download_report.rb]: It isn't immediately obvious to me what `last_object` is until you trace it back up through `classify_download` and back into the each loop at line 40. Is there a way we can clarify that a bit more?

File: app/services/user_stats_importer.rb [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    process_statistic_list(view_report.file_set_page_views, FileViewStat)
+  end
+
+  def gather_download_stats
+    process_statistic_list(download_report.file_set_downloads, FileDownloadStat)
+    process_statistic_list(download_report.work_downloads, FileDownloadStat, :get_download_statistic_object)
+  end
+
+  def tally_user_results
+    sorted_users.each do |user|
```
Comment (awead) [inline app/services/user_stats_importer.rb:31]: where is sorted_users created?

**Review by awead — COMMENTED** [review]
This took me a while to understand, and I'm still not completely sure I follow. I think maybe some yard docs on the methods would help, just a basic what are the inputs and what are the outputs. I found particularly hard to chase down all the instance variables. I left some comments that were my attempts to try and clarify this a bit, but they might be incorrect. 

I think this is okay to merge, but maybe we could try a refactoring session later? I feel like there may be a statistic abstraction buried in here that would really help clear things up.

File: app/services/analytics/view_report.rb (hunk 6 above)
Comment (carolyncole) [reply app/services/analytics/view_report.rb]: @awead I do not know.  I had similar code to that and I removed it becuase the top method seemed silly.

File: app/services/analytics/download_report.rb (hunk 7 above)
Comment (carolyncole) [reply app/services/analytics/download_report.rb]: No downloads is both File and Work downloads.  We could run two queries and just look for files the first time and Works the second time.

File: app/services/analytics/download_report.rb (hunk 8 above)
Comment (carolyncole) [reply app/services/analytics/download_report.rb]: @awead no same as above how could the same method return two different things?

File: app/services/user_stats_importer.rb (hunk 12 above)
Comment (carolyncole) [reply app/services/user_stats_importer.rb:31]: @awead that is a part of the parent class.

Comment (awead) [reply app/services/user_stats_importer.rb:31]: @cam156 I think that needs to be documented then.

File: spec/services/download_report_spec.rb (hunk 3 above)
Comment (cam156) [reply spec/services/download_report_spec.rb]: https://github.com/psu-stewardship/scholarsphere/pull/1028/files#diff-d4425a6b2765b7eb294f239cf6707a6dR42

File: app/services/analytics/download_report.rb (hunk 7 above)
Comment (awead) [reply app/services/analytics/download_report.rb]: That makes more sense to me.

File: spec/services/download_report_spec.rb (hunk 4 above)
Comment (carolyncole) [reply spec/services/download_report_spec.rb]: @awead the code does look for the exception here: https://github.com/psu-stewardship/scholarsphere/pull/1028/files#diff-366be95a4065da0fa459fd2ae2abaa74R52

File: spec/services/download_report_spec.rb (hunk 2 above)
Comment (carolyncole) [reply spec/services/download_report_spec.rb]: @awead I will have to look. They may be left over from when all the tests were together.

File: app/services/analytics/download_report.rb (hunk 7 above)
Comment (carolyncole) [reply app/services/analytics/download_report.rb]: @awead so you really want to send the query to google twice.  That seems unfortunate.

File: app/services/analytics/download_report.rb (hunk 10 above)
Comment (carolyncole) [reply app/services/analytics/download_report.rb]: @awead I do not understand this comment. The arrays are in the each loop.  They are being passed in and returned.  We could just use the instance variables, but you are arguing against them above.

File: app/services/analytics/download_report.rb (hunk 11 above)
Comment (carolyncole) [reply app/services/analytics/download_report.rb]: @awead the last object was the last object pulled out of fedora.  What would you have the name be?  The last_object_loaded_from_fedora ???

File: app/services/user_stats_importer.rb (hunk 12 above)
Comment (carolyncole) [reply app/services/user_stats_importer.rb:31]: @awead You want to document every use of object oriented inheritance in the code?

Comment (carolyncole) [reply app/services/user_stats_importer.rb:31]: If I am hearing correctly any time an object uses a method from the parent object it should be documented, so every time we use find from active fedora we should add a comment that says this is a method inherited from ActiveFedora and is not defined as part of our model, but part of the AftiveFedora::Base

**Comment by carolyncole** [comment]
@awead Hmm.  I seem to have missed on my Naming.  ViewReport grabs a report of all the views from Google Analytics and parses it.  Download Report does similar for downloads.  Let's start at the top.  What would you name those so you can understand what they do?

**Comment by carolyncole** [comment]
@awead based on your comments I assumed the code was completely off.  I spent the morning doing some refactoring and renaming.  Maybe it makes a bit more sense now?

**Comment by awead** [comment]
@cam156 Yes. I think it exposes things that are in need of refactoring later, but some of those are really coming from the super class. I tried my hand at taming this beast and couldn't do any better, so I'd say let just merge it and move on.

**Pull request merged**